crates/astria-composer/tests/blackbox/helper/mock_geth.rs

```diff
@@ -0,0 +1,192 @@
+//!
+//! At the moment, this mock only supports the `eth_subscribe`
+//! RPC with the parameters `["newPendingTransaction", true]` to allow
+//! subscribing to full new pending transactions using the
+//! [`ethers::providers::Middleware::subscribe_full_pendings_txs`] high
+//! level abstraction.
+//!
+//! The intended use for the mock is to:
+//!
+//! 1. spawn a server with [`MockGeth::spawn`];
+//! 2. establish a websocket connection using its local socket address at [`MockGet::local_addr`],
+//!    for example with [`Provider::<Ws>::connect`];
+//! 3. subscribe to its mocked `eth_subscribe` JSONRPC using
+//!    [`Middleware::subscribe_full_pending_txs`];
+//! 4. push new transactions into the server using [`MockGeth::push_tx`], which will subsequently be
+//!    sent to all suscribers and can be observed by the client.
+//!
+//! # Examples
+//!
+//! ```
+//! # tokio_test::block_on( async {
+//! use std::time::Duration;
+//!
+//! use ethers::{
+//!     providers::{Middleware as _, Provider, StreamExt as _, Ws},
+//!     types::Transaction,
+//! };
+//! use jsonrpsee_ethers::MockGeth;
+//!
+//! println!("connecting!!");
+//! let mock_geth = MockGeth::spawn().await;
+//! let server_addr = mock_geth.local_addr();
+//!
+//! tokio::spawn(async move {
+//!     loop {
+//! #       // FIXME: remove the sleep. at the moment this doc tests only passes when
+//! #       //        when expclitly sleeping. Why?
+//!         tokio::time::sleep(Duration::from_secs(1)).await;
+//!         let r = mock_geth.push_tx(Transaction::default());
+//!     }
+//! })
+//!
+//! let geth_client = Provider::<Ws>::connect(format!("ws://{server_addr}"))
+//!     .await
+//!     .expect("client should be able to conenct to local ws server");
+//! let mut new_txs = geth_client.subscribe_full_pending_txs()
+//!     .await
+//!     .unwrap()
+//!     .take(3);
+//! while let Some(new_tx) = new_txs.next().await {
+//!     assert_eq!(new_tx, Transaction::default());
+//! }
+//! # });
+//! ```
+
+use std::net::SocketAddr;
+
+use ethers::types::Transaction;
+use jsonrpsee::{
+    core::{
+        async_trait,
+        SubscriptionResult,
+    },
+    proc_macros::rpc,
+    server::IdProvider,
+    types::{
+        ErrorObjectOwned,
+        SubscriptionId,
+    },
+    PendingSubscriptionSink,
+};
+use tokio::sync::broadcast::{
+    channel,
+    error::SendError,
+    Sender,
+};
+
+#[derive(Debug)]
+pub struct RandomU256IdProvider;
+
+impl IdProvider for RandomU256IdProvider {
+    fn next_id(&self) -> SubscriptionId<'static> {
+        use ethers::types::U256;
+        use impl_serde::serialize::to_hex;
+        use rand::RngCore as _;
+
+        let mut rng = rand::thread_rng();
+        let mut raw_u256 = [0u8; 32];
+        (&mut rng).fill_bytes(&mut raw_u256);
+        // Just in case, convert to u256 and back to big endian because parity's u256
+        // implementation does some extra complex transformations.
```

Comment:
omg what are they doing lmao

Reply:
Basically this: https://github.com/paritytech/parity-common/blob/d3a9327124a66e52ca1114bb8640c02c18c134b8/uint/src/uint.rs#L1240

Not crazy different, but didn't want to worry about something going awry.